tfx/extensions/google_cloud_kubernetes/runner.py

```diff
@@ -0,0 +1,245 @@
+# Copyright 2020 Google LLC. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Helper class to start TFX multi-worker training jobs on GKE."""
+
+import json
+from typing import Any, Dict, List, Text
+
+from absl import logging
+
+from tfx import types
+from tfx import version
+from tfx.types import artifact_utils
+from tfx.utils import kube_utils
+from kubernetes.client.rest import ApiException
+import kubernetes.client as client
+
+# Default TFX container image to use in GKE Training. For GPU training,
+# specify a custom image in executor.TRAINING_ARGS_KEY.
+_TFX_IMAGE = 'tensorflow/tfx:%s' % (version.__version__)
+
+_COMMAND = ["python", "-m", "tfx.scripts.run_executor"]
+
+
```

Comment:
Throughout this file, it would be better if `unique_id` could be replaced with something more descriptive of what unique object it represents.

Reply:
I think the question here is how to create a unique selector / name for the training pods and the services in each pipeline run such that the trainer component would be able to locate them in the cluster. I approached this by taking the unique id of the execution context for the component, or randomly generating one if there is none. Since it was used for both the worker and the service (the one attached to the worker for collective communication), I'm more leaning towards adding some comments instead.